### PR TITLE
Implement XdgDecorationManager 

### PR DESCRIPTION
**Pull request by wineee**
Stack layout prefer csd
Tile layout prefer ssd (no decoration now)

### CODE REVIEW

File: src/server/qtquick/private/wquickxdgdecorationmanager.cpp [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+    W_DECLARE_PUBLIC(WQuickXdgDecorationManager)
+
+    QWXdgDecorationManagerV1 *manager = nullptr;
+    WQuickXdgDecorationManager::DecorationMode m_mode = WQuickXdgDecorationManager::PreferClientSide;
+};
+
+void WQuickXdgDecorationManagerPrivate::onNewToplevelDecoration(QWXdgToplevelDecorationV1 *decorat)
+{
+
```
Comment (zccrs) [inline src/server/qtquick/private/wquickxdgdecorationmanager.cpp:37]: remove

File: src/server/qtquick/private/wquickxdgdecorationmanager.cpp [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    // begin slot function
+    void onNewToplevelDecoration(QWXdgToplevelDecorationV1 *decorat);
+    // end slot function
+
+    W_DECLARE_PUBLIC(WQuickXdgDecorationManager)
+
+    QWXdgDecorationManagerV1 *manager = nullptr;
+    WQuickXdgDecorationManager::DecorationMode m_mode = WQuickXdgDecorationManager::PreferClientSide;
```
Comment (zccrs) [inline src/server/qtquick/private/wquickxdgdecorationmanager.cpp:32]: Don't start with `m_` for variant name of Private class

File: src/server/qtquick/private/wquickxdgdecorationmanager.cpp [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+    if (d->m_mode == mode)
+        return;
+    d->m_mode = mode;
+
+    // notify all decoration that mode changed
+    wlr_xdg_decoration_manager_v1 *wlr_manager = d->manager->handle();
+    wlr_xdg_toplevel_decoration_v1 *wlr_decorations;
+    wl_list_for_each(wlr_decorations, &wlr_manager->decorations, link) {
+        wl_signal_emit_mutable(&wlr_decorations->events.request_mode, &wlr_decorations);
```
Comment (zccrs) [inline src/server/qtquick/private/wquickxdgdecorationmanager.cpp:84]: Why emit the signal in here?

Comment (wineee) [reply src/server/qtquick/private/wquickxdgdecorationmanager.cpp:84]: when we change layout，for example Stack -> Tile,  not only newly opened clients use ssd, we also need to notify all existing clients to reconfigure decoration mode (hide csd)

File: src/server/qtquick/private/wquickxdgdecorationmanager_p.h [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+    enum DecorationMode {
+        DecidesByClient,
+        PreferClientSide,
+        PreferServerSide,
+    };
+    Q_ENUM(DecorationMode)
+
+    void setMode(DecorationMode mode);
+    DecorationMode getMode();
```
Comment (zccrs) [inline src/server/qtquick/private/wquickxdgdecorationmanager_p.h:32]: `DecorationMode mode() const`

File: src/server/qtquick/private/wquickxdgdecorationmanager_p.h [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+#include <QQmlEngine>
+
+WAYLIB_SERVER_BEGIN_NAMESPACE
+
+class WQuickXdgDecorationManagerPrivate;
+class WAYLIB_SERVER_EXPORT WQuickXdgDecorationManager : public WQuickWaylandServerInterface, public WObject
+{
+    Q_OBJECT
+    W_DECLARE_PRIVATE(WQuickXdgDecorationManager)
+    Q_PROPERTY(DecorationMode mode READ getMode WRITE setMode)
```
Comment (zccrs) [inline src/server/qtquick/private/wquickxdgdecorationmanager_p.h:18]: add signal

**Review by zccrs — CHANGES_REQUESTED** [review]

File: src/server/qtquick/private/wquickxdgdecorationmanager.cpp [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    W_Q(WQuickXdgDecorationManager);
+    QObject::connect(decorat,
+                &QWXdgToplevelDecorationV1::requestMode,
+                q,
+                [decorat, this]() {
+                    this->updateDecorationMode(decorat);
+                }
+    );
+    /* For some reason, a lot of clients don't emit the request_mode signal. */
+    Q_EMIT decorat->requestMode();
```
Comment (zccrs) [inline src/server/qtquick/private/wquickxdgdecorationmanager.cpp:47]: Why emit signal? you can call updateDecorationMode at here.

File: src/server/qtquick/private/wquickxdgdecorationmanager.cpp [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+void WQuickXdgDecorationManager::setMode(DecorationMode mode)
+{
+    W_D(WQuickXdgDecorationManager);
+
+    if (d->mode == mode)
+        return;
+    d->mode = mode;
+
+    // update all existing decoration that mode changed
+    wlr_xdg_decoration_manager_v1 *wlr_manager = d->manager->handle();
```
Comment (zccrs) [inline src/server/qtquick/private/wquickxdgdecorationmanager.cpp:86]: the `d->manager` create at `WQuickXdgDecorationManager::create`, maybe its nullptr now.

**Review by zccrs — CHANGES_REQUESTED** [review]

**Review by zccrs — APPROVED** [review]

**Pull request merged**